---
output: html_document
---
#Workflow

*Author: Valeria Velasquez Zapata*

## Part I

First, we need to read the files by using read.file and specifying that it is a tab separated file
```{r}
genotypes <- read.csv("fang_et_al_genotypes.txt", sep="\t")
snp <- read.csv("snp_position.txt", sep="\t")
```

For inpecting data we can use commands as str, nrow, ncol and dim 
```{r}
str(genotypes)
nrow(genotypes)
ncol(genotypes)
dim(genotypes)
```
```{r}
str(snp)
nrow(snp)
ncol(snp)
dim(snp)
```


To know the size of the files in bytes we can use file.size
```{r}
file.size("fang_et_al_genotypes.txt")
file.size("snp_position.txt")
```


To separate the data of teosinte and tripsacum we can use a mathing expression for the groups that constitute each type, named accordingly
```{r}
tripsacum <- genotypes[genotypes$Group=="ZMMIL" | genotypes$Group=="ZMMLR" | genotypes$Group=="ZMMMR", ]
teosinte <- genotypes[genotypes$Group=="ZMPBA" | genotypes$Group=="ZMPIL" | genotypes$Group=="ZMPJA", ]
```

Now we can create the dataframe with the desired columns from snp

```{r}
snp_id_chr_pos <-snp[,c(1,3,4)]
```

After this step we can transpose the files, to get the information of each SNP in one column

```{r}
transposed_teosinte <- as.data.frame(t(teosinte[,-1]))
colnames(transposed_teosinte) <- teosinte$Sample_ID
transposed_tripsacum <- as.data.frame(t(tripsacum[,-1]))
colnames(transposed_tripsacum) <- tripsacum$Sample_ID
```

In order to get a file with all the required columns, we join the transposed files with the snp_id_chr_pos dataframe

```{r}
teosinte_joint <- merge(x = snp_id_chr_pos, y = transposed_teosinte, by.x = "SNP_ID", by.y ="row.names", all.y = TRUE)

tripsacum_joint <- merge(x = snp_id_chr_pos, y = transposed_tripsacum, by.x = "SNP_ID", by.y ="row.names", all.y = TRUE)

```

After this, we can remove the rows that lack of information for chromosome and position:
```{r}
teosinte_joint$Position <- as.numeric(as.character(teosinte_joint$Position))
teosinte_joint$Chromosome <- as.numeric(as.character(teosinte_joint$Chromosome))

teosinte_joint <- teosinte_joint[!(teosinte_joint$Position== "unknown") | !(teosinte_joint$Chromosome== "unknown") | !(teosinte_joint$Chromosome== "multiple"),]
teosinte_joint <- teosinte_joint[!(is.na(teosinte_joint$Position)),]

tripsacum_joint$Position <- as.numeric(as.character(tripsacum_joint$Position))
tripsacum_joint$Chromosome <- as.numeric(as.character(tripsacum_joint$Chromosome))

tripsacum_joint <- tripsacum_joint[!(tripsacum_joint$Position== "unknown") | !(tripsacum_joint$Chromosome== "unknown") | !(tripsacum_joint$Chromosome== "multiple"),]
tripsacum_joint <- tripsacum_joint[!(is.na(tripsacum_joint$Position)),]
```


Now we can sort the files by position:

```{r}

teosinte_joint <- teosinte_joint[order(teosinte_joint$Position),]
tripsacum_joint <- tripsacum_joint[order(tripsacum_joint$Position),]
```

After this, we can split the dataframes according to the chromosome number and save the corresponding files

```{r}
teosinte_split <- split(teosinte_joint, teosinte_joint$Chromosome)
allNames <- names(teosinte_split)
 for(thisName in allNames){
     saveName = paste0('Chr_', thisName, '_teosinte.txt')
     write.table(teosinte_split[[thisName]], file = saveName, quote = FALSE, sep="\t", row.names = FALSE)
 }

tripsacum_split <- split(tripsacum_joint, tripsacum_joint$Chromosome)
allNames <- names(tripsacum_split)
 for(thisName in allNames){
     saveName = paste0('Chr_', thisName, '_tripsacum.txt')
     write.table(tripsacum_split[[thisName]], file = saveName, quote = FALSE, sep="\t", row.names = FALSE)
 }

```

At this point we already have the files with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?

To generate the files with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: - we sort the join files in reverse order and replace the ? symbol per -, using sed

```{r}
teosinte_joint_rev <- teosinte_joint[order(teosinte_joint$Position, decreasing=T),]
tripsacum_joint_rev <- tripsacum_joint[order(tripsacum_joint$Position, decreasing=T),]

```

now we replace the values

```{r}
teosinte_joint_rev[] <- lapply(teosinte_joint_rev, as.character)
teosinte_joint_rev[teosinte_joint_rev == '?/?'] <- '-/-'

tripsacum_joint_rev[] <- lapply(tripsacum_joint_rev, as.character)
tripsacum_joint_rev[tripsacum_joint_rev == '?/?'] <- '-/-'
```

Finally We create all the files


```{r}
teosinte_split_rev <- split(teosinte_joint_rev, teosinte_joint_rev$Chromosome)
allNames <- names(teosinte_split_rev)
 for(thisName in allNames){
     saveName = paste0('Chr_', thisName, '_teosinte_rev.txt')
     write.table(teosinte_split_rev[[thisName]], file = saveName, quote = FALSE, sep="\t", row.names = FALSE)
 }

tripsacum_split_rev <- split(tripsacum_joint_rev, tripsacum_joint_rev$Chromosome)
allNames <- names(tripsacum_split_rev)
 for(thisName in allNames){
     saveName = paste0('Chr_', thisName, '_tripsacum_rev.txt')
     write.table(tripsacum_split_rev[[thisName]], file = saveName, quote = FALSE, sep="\t", row.names = FALSE)
 }
```


**The files named as `Chr_#_teosinte.txt` and `Chr_#_tripsacum.txt` have the SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?**

**The files named as `Chr_#_teosinte_rev.txt` and `Chr_#_tripsacum_rev.txt` have the SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -**

## Part II

In this part we will visualize the data in different ways. To start we will retake the joint dataframes without removing the missing data and NA values, so we recreate the joint files, using a previous command

```{r}
transposed_genotypes <- as.data.frame(t(genotypes[,-1]))
colnames(transposed_genotypes) <- genotypes$Sample_ID

genotypes_joint <- merge(x = snp_id_chr_pos, y = transposed_genotypes, by.x = "SNP_ID", by.y ="row.names", all.y = TRUE)
```

Before starting, we should intall the `reshape2` and `ggplot` package

```{r}
if (!require("reshape2")) install.packages("reshape2")
library("reshape2")
if (!require("ggplot2")) install.packages("ggplot2")
library(ggplot2)
if (!require("plyr")) install.packages("plyr")
library(plyr)
```

The first graphic we will make is a plotting of the total number of SNPs in our dataset on each chromosome. 
```{r}
genotypes_joint$Chromosome <- factor(genotypes_joint$Chromosome, levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown", "NA"))

ggplot(genotypes_joint, aes((Chromosome))) + geom_bar()
```


Then, we can plot the SNPs according to the groups they belong, using the genotypes dataframe. That will tell us What groups contribute most of these SNPs?

```{r}
ggplot(genotypes, aes(Group)) +
  geom_bar()
```

Thne ZMMIL, ZMMLR and ZMPBA contribute the most to the SNP count.

to analyze the amount of heterozygicity we melt the datasets to make them tidy, using a vector with the names of all the SNPs

```{r}
headers_names<- colnames(genotypes)[-c(1:3)]
genotypes_melted <- melt(genotypes, measure.vars = headers_names)
```

After that, we create a new column to indicate whether a particular site is homozygous (has the same nucleotide on both chromosomes (i.e., A/A, C/C, G/G, T/T) or heterozygous (otherwise)). First we assign all missing values as NA: 

```{r}
genotypes_melted[ genotypes_melted == "?/?" ] = NA
genotypes_melted$isHomozygous <- (genotypes_melted$value=="A/A" | genotypes_melted$value=="C/C" | genotypes_melted$value=="G/G" | genotypes_melted$value=="T/T")
```


Then, we sort the dataframe using Group and Species_ID values. 

```{r}
genotypes_sorted_by_ID <- genotypes_melted[order(genotypes_melted$Sample_ID),]
genotypes_sorted_by_Group <- genotypes_melted[order(genotypes_melted$Group),]
```

Finally we make a graph that shows the proportion of homozygous and heterozygous sites as well as missing data in each species (you won't be able to see species names). For doing that, we first built a new dataframe with all the counting values (for homozygous, heterozygous and NA) per Sample_ID, and then we melt the results

```{r}
counting_ID <- ddply(genotypes_sorted_by_ID, c("Sample_ID"), summarise, counting_homozygous=sum(isHomozygous, na.rm=TRUE), counting_heterozygous=sum(!isHomozygous, na.rm=TRUE), isNA=sum(is.na(isHomozygous)))

counting_ID_melt <- melt(counting_ID, measure.vars = c("counting_homozygous", "counting_heterozygous", "isNA"))

ggplot(counting_ID_melt,aes(x = Sample_ID, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack")
```

The plot shows how the homozygous counting is bigger through all SNPs, having low conting for missing data. There are also a group of SNPs that seem not to have heterozygous alleles.

The same process was made when the data is sorted according to the groups:

```{r}
counting_Group <- ddply(genotypes_sorted_by_Group, c("Group"), summarise, counting_homozygous=sum(isHomozygous, na.rm=TRUE), counting_heterozygous=sum(!isHomozygous, na.rm=TRUE), isNA=sum(is.na(isHomozygous)))

counting_Group_melt <- melt(counting_Group, measure.vars = c("counting_homozygous", "counting_heterozygous", "isNA"))

ggplot(counting_Group_melt,aes(x = Group, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack")

```

The last graphic shows how the groups that contribute the most to the SNP number also contribute to the number of heterozygous and homozygous.

Finally, we proposed a visualization of the data based on the calculations of the observed heterozygocity per locus. The general idea is to calculate the observed heterozygocity per SNP as the rate between the number of heterozygous divided by the total of genotyped individuals. After that, we will construch  scatter plot. First, we need to work melt the `genotypes_joint` file, using as measure variables all the genotyped individuals:

```{r}
headers_names_joint<- colnames(genotypes_joint)[-c(1:3)]
genotypes_melted_joint <- melt(genotypes_joint, measure.vars = headers_names_joint)
```
After having the file melted, we can calculate the number of heterozygous, according to the SNP_ID:

```{r}
genotypes_melted_joint[ genotypes_melted_joint == "?/?" ] = NA

genotypes_melted_joint$isHomozygous <- (genotypes_melted_joint$value=="A/A" | genotypes_melted_joint$value=="C/C" | genotypes_melted_joint$value=="G/G" | genotypes_melted_joint$value=="T/T")

genotypes_sorted_by_SNP <- genotypes_melted_joint[order(genotypes_melted_joint$SNP_ID),]

```

After this, we can calculate the observed heterozygosity per SNP and make our plot

```{r}
Observed_Het_per_locus <- ddply(genotypes_sorted_by_SNP, c("SNP_ID"), summarise, heterozygocity_count=sum(!isHomozygous, na.rm=TRUE), total_count=sum(!is.na(isHomozygous)))

Observed_Het_per_locus$Obs_heterozygocity <- (Observed_Het_per_locus$heterozygocity_count/Observed_Het_per_locus$total_count)

Observed_Het_per_locus_melt <- melt(Observed_Het_per_locus, measure.vars = "Obs_heterozygocity")

ggplot(Observed_Het_per_locus_melt,aes(x = SNP_ID, y= value, fill=variable)) + geom_point()
```

From the last plot it is possible to see the ranges distribution of the observed heterozygocity for each SNP, being able to identify that the majority of the SNP had an observed heterosygocity between 0 and 0,27, and that one of them has a velue of 1.0 

